gnet_test.go

```diff
@@ -411,6 +413,171 @@ func startClient(t *testing.T, network, addr string, multicore, async bool) {
+		// https://www.iana.org/assignments/ipv6-multicast-addresses/ipv6-multicast-addresses.xhtml#link-local
+		t.Run("udp-multicast", func(t *testing.T) {
+			testMulticast(t, fmt.Sprintf("[ff02::3%%%s]:9991", iface.Name), false, false, iface.Index, 10)
+		})
+		t.Run("udp-multicast-reuseport", func(t *testing.T) {
+			testMulticast(t, fmt.Sprintf("[ff02::3%%%s]:9991", iface.Name), true, false, iface.Index, 10)
+		})
+		t.Run("udp-multicast-reuseaddr", func(t *testing.T) {
+			testMulticast(t, fmt.Sprintf("[ff02::3%%%s]:9991", iface.Name), false, true, iface.Index, 10)
+		})
+	})
+}
+
+func findLoopbackInterface() (*net.Interface, error) {
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+	for _, iface := range ifaces {
+		if iface.Flags&net.FlagLoopback == net.FlagLoopback {
+			return &iface, nil
+		}
+	}
+	return nil, errors.New("no loopback interface")
+}
+
+func testMulticast(t *testing.T, addr string, reuseport, reuseaddr bool, index, nclients int) {
+	ts := &testMcastServer{
+		t:        t,
+		addr:     addr,
+		nclients: nclients,
+	}
+	options := []Option{
+		WithReuseAddr(reuseaddr),
+		WithReusePort(reuseport),
+		WithSocketRecvBuffer(2 * nclients * 1024), // enough space to receive messages from nclients to eliminate dropped packets
+		WithTicker(true),
+	}
+	if index != -1 {
+		options = append(options, WithMulticastInterfaceIndex(index))
+	}
+	err := Run(ts, "udp://"+addr, options...)
+	assert.NoError(t, err)
+}
+
+type testMcastServer struct {
+	*BuiltinEventEngine
+	t        *testing.T
+	mcast    sync.Map
+	addr     string
+	nclients int
+	started  int32
+	active   int32
+}
+
+func (s *testMcastServer) startMcastClient() {
+	rand.Seed(time.Now().UnixNano())
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	c, err := net.Dial("udp", s.addr)
+	require.NoError(s.t, err)
+	defer c.Close()
+	ch := make(chan []byte, 10000)
+	s.mcast.Store(c.LocalAddr().String(), ch)
+	duration := time.Duration((rand.Float64()*2+1)*float64(time.Second)) / 2
+	s.t.Logf("test duration: %dms", duration/time.Millisecond)
+	start := time.Now()
+	for time.Since(start) < duration {
+		reqData := make([]byte, 1024)
+		_, err = rand.Read(reqData)
+		require.NoError(s.t, err)
+		_, err = c.Write(reqData)
+		require.NoError(s.t, err)
+		// Workaround for MacOS "write: no buffer space available" error messages
+		// https://developer.apple.com/forums/thread/42334
+		time.Sleep(time.Millisecond)
+		select {
+		case respData := <-ch:
+			require.Equalf(s.t, reqData, respData, "response mismatch, length of bytes: %d vs %d", len(reqData), len(respData))
+		case <-ctx.Done():
+			require.Fail(s.t, "timeout receiving message")
+			return
+		}
+	}
+}
+
+func (s *testMcastServer) OnTraffic(c Conn) (action Action) {
+	buf, _ := c.Next(-1)
+	b := make([]byte, len(buf))
+	copy(b, buf)
+	ch, ok := s.mcast.Load(c.RemoteAddr().String())
+	require.True(s.t, ok)
+	ch.(chan []byte) <- b
```

Comment:
Why do we need this `ch`? Why don't you just call `c.Write(b)`?

Reply:
I think it is not possible to send a message back from a multicast address. If we try to, then we get `can't assign requested address` error on Mac, or the message will return with a different source address on Linux.